### PR TITLE
TestClusterTopology: support setting labels for different tikv instances

### LINKED ISSUE #108: Support different configuration for different nodes
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

There are some configuration items like labels for tikv and location-lables for pd, which can be varied on different nodes in the same cluster. Since tiup already supports this, we need to enhance naglfar to support this as well.

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by CadmusJiang**
…l label by config

Signed-off-by: Cadmus <CadmusJiang@gmail.com>

### What problem does this PR solve?

Issue Number: close #108  

Problem Summary: set label for tikv

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works: 
1.add config field in tikv, tidb, pd
2.modify the original check immutable webhook to allow the update of config

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)
```yaml
  serverConfigs:
      ...
      pd: |-
     ...
        replication.location-labels: ["zone","dc","rack","host"]
    ....
    tikv:
      ....
        config: |-
          server.labels:
            zone: bj
            dc: bja
            rack: rack1
            host: host2
```
1.create testclustertopology include above configuration
2.observe the log check effect
tips: replication.location-labels must be set when the cluster is created
Side effects

- Breaking backward compatibility

### CODE REVIEW

File: controllers/testclustertopology_controller.go [hunk 1]
```diff
@@ -626,11 +626,11 @@ func (r *TestClusterTopologyReconciler) updateTiDBCluster(ctx context.Context, c
 			return false, ctrl.Result{RequeueAfter: 10 * time.Second}, nil
 		}
 	}
-	if cluster.IsServerConfigModified(ct.Status.PreTiDBCluster.ServerConfigs, ct.Spec.TiDBCluster.ServerConfigs) {
+	if cluster.IsServerConfigModified(ct.Status.PreTiDBCluster.ServerConfigs, ct.Spec.TiDBCluster.ServerConfigs) || cluster.IsComponentsConfigModified(ct.Status.PreTiDBCluster, ct.Spec.TiDBCluster) {
```
Comment (mahjonp) [inline controllers/testclustertopology_controller.go:629]: What's the difference between `cluster.IsComponentsConfigModified` and `checkImmutableFieldChanged`? IMO you could retain only one of them.

**Review by mahjonp — COMMENTED** [review]
Rest LGTM

File: api/v1/testclustertopology_webhook.go [hunk 2]
```diff
@@ -320,17 +301,91 @@ func checkVersionDownloadURL(pre *TiDBCluster, cur *TiDBCluster) bool {
...
 	updatedModules := 0
+	var modules []string
 	if checkScale(pre, cur) {
 		updatedModules++
+		modules = append(modules, "scale-in/out")
 	}
-	if checkServerConfigModified(&pre.ServerConfigs, &cur.ServerConfigs) {
+	if checkServerConfigModified(&pre.ServerConfigs, &cur.ServerConfigs) || checkComponentsConfigModified(pre, cur) {
 		updatedModules++
+		modules = append(modules, "update")
```
[the remark this answers is not in the document]
Comment (CadmusJiang) [reply api/v1/testclustertopology_webhook.go:382]: done

File: api/v1/testclustertopology_webhook.go [hunk 3]
```diff
@@ -320,17 +301,91 @@ func checkVersionDownloadURL(pre *TiDBCluster, cur *TiDBCluster) bool {
 	return pre.Version.PDDownloadURL != cur.Version.PDDownloadURL || pre.Version.TiDBDownloadURL != cur.Version.TiDBDownloadURL || pre.Version.TiKVDownloadURL != cur.Version.TiKVDownloadURL
 }
 
+// checkImmutableFieldChanged check if immutable fields are changed, like spec.tidbCluster.tidb[i].dataDir
+func checkImmutableFieldChanged(pre *TiDBCluster, cur *TiDBCluster) (bool, error) {
```
[the remark this answers is not in the document]
Comment (CadmusJiang) [reply api/v1/testclustertopology_webhook.go:305]: done

File: api/v1/testclustertopology_webhook.go [hunk 4]
```diff
@@ -125,17 +127,21 @@ func (r *TestClusterTopology) validateTiDBUpdate(tct *TestClusterTopology) error
 		return fmt.Errorf("upgradePolicy must be `force` or empty")
 	}
 
+	isChanged, err := checkImmutableFieldChanged(tct.Status.PreTiDBCluster, r.Spec.TiDBCluster)
```
[the remark this answers is not in the document]
Comment (CadmusJiang) [reply api/v1/testclustertopology_webhook.go:130]: done

File: controllers/testclustertopology_controller.go (hunk 1 above)
Comment (CadmusJiang) [reply controllers/testclustertopology_controller.go:629]: done

**Comment by CadmusJiang** [comment]
5421614675938_ pic_hd
go-cmp Diff return string, it is not suitable to program.

File: api/v1/testclustertopology_webhook.go [hunk 5]
```diff
@@ -312,28 +290,111 @@ func countClusterNum(tct *TestClusterTopology) int {
...
+			continue
+		}
+		for j := 0; j < preComponents.Len(); j++ {
+			for k := 0; k < curComponents.Len(); k++ {
+				if preComponents.Index(j).FieldByName("Host").String() == curComponents.Index(k).FieldByName("Host").String() {
+					preComponent := preComponents.Index(j).Interface()
+					curComponent := curComponents.Index(k).Interface()
+
+					differ, err := diff.NewDiffer(
+						diff.TagName("{ Project: Naglfar }"),
```
Comment (mahjonp) [inline api/v1/testclustertopology_webhook.go:322]: test please.

**Review by mahjonp — APPROVED** [review]

**Pull request merged**